Implement transform feedback (GX2*StreamOut* functions).

achurch opened the pull request:
See issue #234.

exjam commented on src/libdecaf/src/modules/gx2/gx2_shaders.cpp:
```diff
@@ -579,6 +601,54 @@ GX2SetStreamOutEnable(BOOL enable)
+   case GX2StreamOutContextMode::FromOffset:
+      control = control
+         .OFFSET_SOURCE().set(pm4::STRMOUT_OFFSET_FROM_PACKET);
+      srcLo = 0u;  // TODO: where does the offset come from?
```
BTW it definitely does reuse stream here as the offset, just do srcLo = mem::untranslate(stream)

achurch replied:
Hmm, if that's the case then the offset mode needs to be something else, because FROM_PACKET would try to use the memory address as an offset into the buffer (which is of course going to overflow). Can you doublecheck, or maybe put up the disassembly somewhere so I can take a look at it?

It could of course be a bug in GX2 itself...

exjam commented on src/libdecaf/src/gpu/pm4.h:
```diff
@@ -718,6 +718,54 @@ struct EventWriteEOP
+BITFIELD(SBU_CONTROL, uint32_t)
+   BITFIELD_ENTRY(0, 1, uint32_t, STORE_BUFFER_FILLED_SIZE);
+   BITFIELD_ENTRY(1, 2, STRMOUT_OFFSET_SOURCE, OFFSET_SOURCE);
+   BITFIELD_ENTRY(8, 2, bool, SELECT_BUFFER);
```
This is a 2 bit field, change type from bool to uint8_t

achurch replied:
buuuh, I is stupid.
Also made STORE a bool while I was at it.

exjam commented:
Also could you rebase on top of this change I just pushed to master d92925d6622724dfc37247f34b40eea234049a28 ?

Fairly simple change, it just changes bitfields so:
bitfield.field().get() becomes bitfield.field()
bitfield.field().set(x) becomes bitfield.field(x)

achurch commented:
Rebase done. I haven't actually tried building it yet, will look at the build output and fix anything that needs it.

exjam commented:
Can't we just map a stream out context to an OpenGL transform feedback object so we can avoid the whole glBeginQuery, glBeginTransformFeedback, glEndTransformFeedback, glEndQuery, glGetQueryObjectuiv around every single draw call? Because this series of functions is forcing the cpu & gpu to sync in order to get the query result right?

achurch commented:
Rebased again, and added a cache on the buffer enable bits.

Re transform feedback objects, from what I read they're not a panacea like that -- you still need the Begin/EndTransformFeedback in any case, and XFB objects are only good for caching draw counts if you reuse the count on the GPU side -- you still need a query to retrieve it on the CPU side.

That said, there are probably some cases we could detect and optimize where the client code passes the feedback buffer to a shader, and in that case XFB objects would be a win.
